Task 6

IvanPavlov90 opened the pull request:
Task 6.1
https://htmlpreview.github.io/?https://github.com/IvanPavlov90/EPAM-XT-2021/blob/Task_6/Task%206/Task%206.1/index.html

Task 6.2
https://htmlpreview.github.io/?https://github.com/IvanPavlov90/EPAM-XT-2021/blob/Task_6/Task%206/Task%206.2/index.html

Task 6.3
https://htmlpreview.github.io/?https://github.com/IvanPavlov90/EPAM-XT-2021/blob/Task_6/Task%206/Task%206.3/index.html

Task 6.4
https://htmlpreview.github.io/?https://github.com/IvanPavlov90/EPAM-XT-2021/blob/Task_6/Task%206/Task%206.4/index.html

IvanPavlov90 committed:
finish task 6.1

IvanPavlov90 committed:
create form, styles for inputs and labels.

IvanPavlov90 committed:
finish task 6.2 (beside colors)

IvanPavlov90 committed:
tasks 6.1 and 6.3

- fix menu width (6.1);
- create markup for 6.3 task;
- create triangles;

IvanPavlov90 committed:
fix colours in tasks 6.2 and 6.3

IvanPavlov90 committed:
finish task 6.3 and 6.4

IvanPavlov90 committed:
add box-shadow for 6.4 and little width fix in 6.3

IvanPavlov90 committed:
fix with file style.css in preview

IvanPavlov90 commented:
Task 6.1 
https://htmlpreview.github.io/?https://github.com/IvanPavlov90/EPAM-XT-2021/blob/Task_6/Task%206/Task%206.1/index.html

IvanPavlov90 commented:
Task 6.2
https://htmlpreview.github.io/?https://github.com/IvanPavlov90/EPAM-XT-2021/blob/Task_6/Task%206/Task%206.2/index.html

Task 6.3
https://htmlpreview.github.io/?https://github.com/IvanPavlov90/EPAM-XT-2021/blob/Task_6/Task%206/Task%206.3/index.html

Task 6.4
https://github.com/IvanPavlov90/EPAM-XT-2021/blob/Task_6/Task%206/Task%206.4/index.html

STeeL835 commented on Task 6/Task 6.1/index.html:
```diff
@@ -0,0 +1,18 @@
+<html lang="ru" dir="ltr">
+  <head>
+    <meta charset="utf-8">
+    <meta name = "viewport" content = "width = device-width, initial-scale = 1.0">
```
in web attributes don't use such spacing - there can be a lot of attributes and this will only make the line longer. It's okay if equals sign doesn't have spaces around.

STeeL835 commented on Task 6/Task 6.1/style.css:
```diff
@@ -0,0 +1,24 @@
+.mainList li {
+  margin-left: -20%;
+  margin-top: 0.5rem;
+}
```
<details><summary>Image - Real list position</summary>

![изображение](https://user-images.githubusercontent.com/25896927/119192301-f1d11300-ba90-11eb-932a-1be83cb65ba8.png)
</details>

You are fighting with the consequences instead of reason. You should've adjusted `ul` padding instead of making negative margin for list items.

Also, visible paddings aren't even - right side isn't prepared for text that has just the right width

STeeL835 commented on Task 6/Task 6.1/style.css:
```diff
@@ -0,0 +1,24 @@
+  text-align: left;
+}
+
+.mainList li {
```
If you're making this style specifically for lists - use more precise selectors:
```css
ul.mainList,
ol.mainList {
/* ... */
ul.mainList > li,
ol.mainList > li {
```
This style can look ugly on other elements. But if you are sure that you want to make the style for all elements, then okay - but here I see that the second style can be applied to `li` that is 4 levels deeper than element with this class, which can be unexpected

STeeL835 commented on Task 6/Task 6.1/style.css:
```diff
@@ -0,0 +1,24 @@
+  list-style: none;
+  border: 1px solid black;
+  border-radius: 10px;
+  box-shadow: 5px 5px 0 0px black;
```
Shadow should be brown too

STeeL835 commented on Task 6/Task 6.2/index.html:
```diff
@@ -0,0 +1,33 @@
+</head>
+<body>
+    <form action="" class = "mainForm">
+        <p class = "formHeader">Регистрация</p>
```
`p` element isn't used for headers. There are `header`/`h123456` elements for this. This would also allow you to implicitly style them (without adding classes for them, but with writing rules for them using element names)

Also, usually for such forms `fieldset` tag is used with `legend` tag as header

STeeL835 commented on Task 6/Task 6.2/style.css:
```diff
@@ -0,0 +1,69 @@
+
+.registrationUnderline {
+    background-color: white;
+}
```
* you could implicitly style it like `form.mainForm hr`
* hr doesn't have background, you should've used `color`

STeeL835 commented on Task 6/Task 6.2/style.css:
```diff
@@ -0,0 +1,69 @@
+    margin: 0% 0% 5% 4%;
+    font-size: 18px;
+    background-color: #ffdf7f;
+}
```
Forgot to remove borders

Also, this definitely could be implicitly styled

STeeL835 commented on Task 6/Task 6.2/index.html:
```diff
@@ -0,0 +1,33 @@
+        <p class = "formHeader">Регистрация</p>
+        <hr class = "registrationUnderline">
+        <div class = "e-mail"> 
+            <label for="e-mail">E-mail</label><br>
```
* dont put 2 tags on the same line
* there's no need in `br` anyway - input has 100% width and won't let itself be on the same line as label

STeeL835 commented on Task 6/Task 6.2/index.html:
```diff
@@ -0,0 +1,33 @@
+    <form action="" class = "mainForm">
+        <p class = "formHeader">Регистрация</p>
+        <hr class = "registrationUnderline">
+        <div class = "e-mail"> 
+            <label for="e-mail">E-mail</label><br>
+            <input type="text" id = "e-mail">
+        </div>
+        <div class = "passwordContainter">
+            <div class = "password">
+                <label for="password">Пароль</label><br>
+                <input type="text" id = "password">
+            </div>
+            <div class = "passwordOnceAgain">
+                <label for="passwordOnceAgain">Ещё раз</label><br>
+                <input type="text" id = "passwordOnceAgain">
+            </div>
```
<details><summary>Image - Real block position</summary>

![изображение](https://user-images.githubusercontent.com/25896927/119195353-90f80980-ba95-11eb-9a6b-7993af7e5338.png)
</details>

You used flexbox AND width 50% AND fixed width for input. And also, left margin. You should've used less - maybe flexbox with `justify-content: space-between` and fixed width or do without flex box, by making `div` an inline-block and adding a margin. Also - you either use 50% or fixed width in this case

STeeL835 replied:
You didn't make the inputs required - I can hit submit without filling these fields

STeeL835 commented on Task 6/Task 6.2/index.html:
```diff
@@ -0,0 +1,33 @@
+            <div class = "password">
+                <label for="password">Пароль</label><br>
+                <input type="text" id = "password">
+            </div>
```
<details><summary>Image - input out of block</summary>

![изображение](https://user-images.githubusercontent.com/25896927/119196874-efbe8280-ba97-11eb-916e-0b8cbfa9e422.png)
</details>

This is because input's `box-sizing` is `content-box` by default - so it calculates width by content part of block model. You can set it to `border-box` to calculate padding and borders too

STeeL835 commented on Task 6/Task 6.3/style.css:
```diff
@@ -0,0 +1,35 @@
+    min-width: 300px;
+}
+
+.mainBlock::before {
+    content: "";
+    border-top: 50px solid transparent;
+    border-left: 50px solid transparent;
+    border-right: 50px solid transparent;
+    border-bottom: 50px solid #c2e1f5;
+    position: absolute;
+    top: -30px;
+    margin-left: -50px;
+}
```
since you don't have `position:relative` in your parent element (element is a parent to its pseudo elements), pseudo elements are positioning relative to a page. If I move the element, triangles won't move after the element.

STeeL835 commented on Task 6/Task 6.3/style.css:
```diff
@@ -0,0 +1,35 @@
+    margin-left: -50px;
+}
+
+.mainBlock::after {
+    content: "";
+    border-top: 46px solid transparent;
+    border-left: 46px solid transparent;
+    border-right: 46px solid transparent;
+    border-bottom: 46px solid #88b7d5;
+    position: absolute;
+    top: -17px;
+    margin-left: -46px;
+}
```
You can use css variables and `calc()` do reduce the amount of magic numbers

STeeL835 commented on Task 6/Task 6.4/index.html:
```diff
@@ -0,0 +1,31 @@
+    <link rel="stylesheet" href="index.css">
+  </head>
+  <body>
+    <section class = "mainSection">
+      <div class="section">
```
you have an element `section` and a class `section` - confusing

STeeL835 commented on Task 6/Task 6.4/index.css:
```diff
@@ -0,0 +1,50 @@
+  border-radius: 2px;
+}
+
+.image {
+  width: 65px;
+  height: 60px;
+}
```
I guess, you can limit only height here. If you are afraid of image being too wide, you can use `max-width` property (with `max-height` if you want to save proportions) - they gently limit image size

STeeL835 commented on Task 6/Task 6.4/index.css:
```diff
@@ -0,0 +1,50 @@
+}
+
+.mainSection div:last-child {
+  border-right: 0px;
```
```suggestion
  border-right: 0px;
```
```suggestion
  border-right: none;
```

STeeL835 commented on Task 6/Task 6.4/index.html:
```diff
@@ -0,0 +1,31 @@
+  <body>
+    <section class = "mainSection">
+      <div class="section">
+        <img src="data-collection.png" alt="Cloud icon" class = "image"></img>
```
* Not a "cloud icon"
* This can be implicitly styled (or class name could be less naive)

STeeL835 reviewed (commented):
Overall I'd suggest making styles more universal - 
* instead of writing a style for every element semantically, try to make it generic (`passwordOnceAgain` vs `half-width-input`)
* make several styles where it makes sense (`class="form-input half"` - they combine)
* don't create class where it makes sense to rely on element name or position in the tree (`.form-input.half > label`, `.form-input.half > input`)
* Use `kebab-case` for css names, not `camelCase`

IvanPavlov90 committed:
style correction for tasks 6.1 and 6.2:

- remove extra spaces after sign "=" in both tasks;
- change margin to padding in task 6.1 and also rewrite ul style;
- change shadow color to brown in task 6.1:
- change <p> to <h1> in header in task 6.2;
- change styles for line and button inside form in 6.2;
- change styles that are related to <div> password-container;
- add required attribute to form fields;

IvanPavlov90 committed:
corrections for tasks 6.3 and 6.4:

- apply relative position to pseudoelements in 6.3 task;
- rename div classes in task 6.4, for now their class names are not equal to section tagname;
- apply some style changes to images in task 6.4;

IvanPavlov90 committed:
add variables into task 6.3

IvanPavlov90 committed:
little fix with input types

IvanPavlov90 committed:
fix 6.2 names

STeeL835 commented on Task 6/Task 6.3/style.css:
```diff
@@ -0,0 +1,44 @@
+    min-width: 350px;
+}
+
+.main-block::before {
+    content: "";
+    border-top: var(--border-width-before-element) solid transparent;
+    border-left: var(--border-width-before-element) solid transparent;
+    border-right: var(--border-width-before-element) solid transparent;
+    border-bottom: var(--border-width-before-element) solid #c2e1f5;
+    position: relative;
+    top: -72px;
+}
+
+.main-block::after {
+    content: "";
+    border-top: var(--border-width-after-element) solid transparent;
+    border-left: var(--border-width-after-element) solid transparent;
+    border-right: var(--border-width-after-element) solid transparent;
+    border-bottom: var(--border-width-after-element) solid #88b7d5;
+    position: relative;
+    top: -124px;
+}
```
You could save on duplication in 2 ways:
* You can write 1 `border` style, and override `border-bottom`
* You can write styles twice, but combining them like:
```css
.main-block::before,
.main-block::after {
 /* shared stuff */
}

.main-block::before { }

.main-block::after { }
```

IvanPavlov90 committed:
change duplications in trinagle styles in 6.3
